### PR TITLE
[FIX] Add ClassVar type to mutable class defaults

### PR DESCRIPTION
**Pull request by badrinini**
his Pull Request fixes all occurrences of the RUF012 error related to the use of mutable default values in Python classes. The RUF012 error occurs when a mutable object (such as a list or dictionary) is used as a default value for a class argument, which can lead to unexpected behavior due to the mutability of these objects.

Changes made:
Replaced mutable default values (e.g., lists, dictionaries) with immutable defaults (e.g., None or immutable objects like tuples).
Updated the code to ensure that default values are safe and do not cause unintended behavior when modified by multiple class instances.
Why these changes?
Using mutable default values in method or class constructors can lead to hard-to-detect bugs. By using immutable default values (such as None), we avoid these issues and adhere to best practices recommended by the ruff tool.

Related Links
[RUF012 - mutable-class-default](https://docs.astral.sh/ruff/rules/#mutable-class-default)

### CODE REVIEW

File: nilearn/regions/parcellations.py [hunk 1]
```diff
@@ -259,14 +261,17 @@ class Parcellations(_MultiPCA):
...
         "ward",
         "complete",
         "average",
         "rena",
         "hierarchical_kmeans",
-    ]
+    )
+
+    def __init__(self):
+        pass
```
Comment (Remi-Gau) [inline nilearn/regions/parcellations.py:274]:
```suggestion
```

I don't think this is needed

**Review by Remi-Gau — CHANGES_REQUESTED** [review]
Can you delete this line from the pyproject.toml to make sure this PR fixes all instances of this error?
https://github.com/nilearn/nilearn/blob/5d287b127aab0247c682c744739a9b67a3d2ed85/pyproject.toml#L167

File: nilearn/decoding/space_net.py [hunk 2]
```diff
@@ -12,6 +12,7 @@
 
 import collections
 import time
+from typing import ClassVar
```
Comment (Remi-Gau) [inline nilearn/decoding/space_net.py:15]: You are getting some code style errors on the import.

See here: https://results.pre-commit.ci/run/github/1235740/1734281156.IFMbzslZRvWBfnsXqlJd_Q

My suggestion to fix them.

Install tox: `pip install tox`

From the root of the repository run:

```
tox
```

This should autofix most issues that you can them commit and push.

File: nilearn/plotting/displays/_slicers.py [hunk 3]
```diff
@@ -46,7 +47,7 @@ class BaseSlicer:
     """
 
     # This actually encodes the figsize for only one axe
-    _default_figsize = [2.2, 2.6]
+    _default_figsize: ClassVar[tuple[float, float]] = (2.2, 2.6)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_slicers.py:50]:
```suggestion
    _default_figsize: ClassVar[list[float, float]] = (2.2, 2.6)
```

File: nilearn/plotting/displays/_slicers.py [hunk 4]
```diff
@@ -898,9 +899,9 @@ class OrthoSlicer(BaseSlicer):
 
     """
 
-    _cut_displayed = "yxz"
+    _cut_displayed: ClassVar[str] = "yxz"
     _axes_class = CutAxes
-    _default_figsize = [2.2, 3.5]
+    _default_figsize: ClassVar[tuple[float, float]] = (2.2, 3.5)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_slicers.py:904]:
```suggestion
    _default_figsize: ClassVar[list[float, float]] = (2.2, 3.5)
```

File: nilearn/plotting/displays/_slicers.py [hunk 5]
```diff
@@ -1132,9 +1133,9 @@ class TiledSlicer(BaseSlicer):
 
     """
 
-    _cut_displayed = "yxz"
+    _cut_displayed: ClassVar[str] = "yxz"
     _axes_class = CutAxes
-    _default_figsize = [2.0, 7.6]
+    _default_figsize: ClassVar[tuple[float, float]] = (2.0, 7.6)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_slicers.py:1138]:
```suggestion
    _default_figsize: ClassVar[list[float, float]] = (2.0, 7.6)
```

File: nilearn/plotting/displays/_slicers.py [hunk 6]
```diff
@@ -1648,8 +1649,8 @@ class XSlicer(BaseStackedSlicer):
 
     """
 
-    _direction = "x"
-    _default_figsize = [2.6, 2.3]
+    _direction: ClassVar[str] = "x"
+    _default_figsize: ClassVar[tuple[float, float]] = (2.6, 2.3)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_slicers.py:1653]:
```suggestion
    _default_figsize: ClassVar[list[float, float]] = (2.6, 2.3)
```

File: nilearn/plotting/displays/_slicers.py [hunk 7]
```diff
@@ -1687,8 +1688,8 @@ class YSlicer(BaseStackedSlicer):
 
     """
 
-    _direction = "y"
-    _default_figsize = [2.2, 3.0]
+    _direction: ClassVar[str] = "y"
+    _default_figsize: ClassVar[tuple[float, float]] = (2.2, 3.0)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_slicers.py:1692]:
```suggestion
    _default_figsize: ClassVar[list[float, float]] = (2.2, 3.0)
```

File: nilearn/plotting/displays/_slicers.py [hunk 8]
```diff
@@ -1726,8 +1727,8 @@ class ZSlicer(BaseStackedSlicer):
 
     """
 
-    _direction = "z"
-    _default_figsize = [2.2, 3.2]
+    _direction: ClassVar[str] = "z"
+    _default_figsize: ClassVar[tuple[float, float]] = (2.2, 3.2)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_slicers.py:1731]:
```suggestion
    _default_figsize: ClassVar[list[float, float]] = (2.2, 3.2)
```

File: nilearn/plotting/displays/_slicers.py [hunk 9]
```diff
@@ -1841,8 +1842,8 @@ class YZSlicer(OrthoSlicer):
 
     """
 
-    _cut_displayed = "yz"
-    _default_figsize = [2.2, 3.0]
+    _cut_displayed: ClassVar[str] = "yz"
+    _default_figsize: ClassVar[tuple[float, float]] = (2.2, 3.0)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_slicers.py:1846]:
```suggestion
    _default_figsize: ClassVar[list[float, float]] = (2.2, 3.0)
```

File: nilearn/plotting/displays/_slicers.py [hunk 10]
```diff
@@ -1884,9 +1885,9 @@ class MosaicSlicer(BaseSlicer):
 
     """
 
-    _cut_displayed = "yxz"
-    _axes_class = CutAxes
-    _default_figsize = [4.0, 5.0]
+    _cut_displayed: ClassVar[str] = "yxz"
+    _axes_class: ClassVar[CutAxes] = CutAxes
+    _default_figsize: ClassVar[tuple[float, float]] = (4.0, 5.0)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_slicers.py:1890]:
```suggestion
    _default_figsize: ClassVar[list[float, float]] = (4.0, 5.0)
```

**Review by Remi-Gau — CHANGES_REQUESTED** [review]
You are missing a few more changes from tuple to list

**Review by Remi-Gau — CHANGES_REQUESTED** [review]
You also need to update the changelog and add yourself as a contributor in the CITATION.cff file in the root of the repo:

see here for more info
https://nilearn.github.io/stable/development.html#changelog

**Comment by badrinini** [comment]
I hope you are doing well.

I wanted to inform you that I have addressed the error you pointed out and
have also implemented the recommendations you provided.

I would be grateful if you could take a moment to review the changes.
Should you have any further suggestions or if anything requires additional
adjustments, please feel free to let me know.

Thank you once again for your valuable guidance and assistance.

Best regards,
Traore Idrissa

File: CITATION.cff [hunk 11]
```diff
@@ -777,3 +777,7 @@ authors:
     family-names: Nájera
     website: https://github.com/Titan-C
     affiliation: Checkmk
+    given-names: IDRISSA
+    family-names: TRAORE
```
Comment (Remi-Gau) [inline CITATION.cff:781]:
```suggestion
    given-names: Idrissa
    family-names: Traore
```

Comment (Remi-Gau) [reply CITATION.cff:781]: let's only have the first letter in upper case

File: doc/changes/latest.rst [hunk 12]
```diff
@@ -1,5 +1,5 @@
 .. currentmodule:: nilearn
-
+:bdg-dark:Code Fix all occurrences of the RUF012 error related to mutable default values in Python classes (:gh:[PR_NUMBER] by TRAORE IDRISSA_).
```
Comment (Remi-Gau) [inline doc/changes/latest.rst:2]:
```suggestion
:bdg-dark:Code Fix all occurrences of the RUF012 error related to mutable default values in Python classes (:gh:[4954] by Idrissa Traore_).
```

File: CITATION.cff [hunk 13]
```diff
@@ -764,3 +764,6 @@ authors:
     family-names: Nájera
     website: https://github.com/Titan-C
     affiliation: Checkmk
+    given-names: Idrissa
+    family-names: Traore
+    affiliation: IMT Atlantique, Brest, France
```
Comment (Remi-Gau) [inline CITATION.cff:769]:
```suggestion
  - given-names: Idrissa
    family-names: Traore
    affiliation: IMT Atlantique, Brest, France
```

**Comment by Remi-Gau** [comment]
@badrinini 
I fixed a couple of issues: some classes attributes were still typed as tuple and there were a few styling issues

**Review by Remi-Gau — APPROVED** [review]
Good with me

**Comment by Remi-Gau** [comment]
oops tests are failing because we updated the type annotation but not the code: let me fix that

File: nilearn/plotting/displays/_projectors.py [hunk 14]
```diff
@@ -322,8 +323,8 @@ class XProjector(OrthoProjector):
 
     """
 
-    _cut_displayed = "x"
-    _default_figsize = [2.6, 3.0]
+    _cut_displayed: ClassVar[str] = "x"
+    _default_figsize: ClassVar[list[float, float]] = (2.6, 3.0)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_projectors.py:327]:
```suggestion
    _default_figsize: ClassVar[list[float, float]] = [2.6, 3.0]
```

File: nilearn/plotting/displays/_projectors.py [hunk 15]
```diff
@@ -356,8 +357,10 @@ class YProjector(OrthoProjector):
 
     """
 
-    _cut_displayed = "y"
-    _default_figsize = [2.2, 3.0]
+    _cut_displayed: ClassVar[str] = "y"
+
+
+_default_figsize: ClassVar[list[float, float]] = (2.2, 3.0)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_projectors.py:363]:
```suggestion
    _cut_displayed: ClassVar[str] = "y"
    _default_figsize: ClassVar[list[float, float]] = [2.2, 3.0]
```

File: nilearn/plotting/displays/_projectors.py [hunk 16]
```diff
@@ -390,8 +393,9 @@ class ZProjector(OrthoProjector):
 
     """
 
-    _cut_displayed = "z"
-    _default_figsize = [2.2, 3.4]
+
+_cut_displayed: ClassVar[str] = "z"
+_default_figsize: ClassVar[list[float, float]] = (2.2, 3.4)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_projectors.py:398]:
```suggestion
_default_figsize: ClassVar[list[float, float]] = [2.2, 3.4]
```

File: nilearn/plotting/displays/_projectors.py [hunk 17]
```diff
@@ -498,8 +502,9 @@ class YZProjector(OrthoProjector):
 
     """
 
-    _cut_displayed = "yz"
-    _default_figsize = [2.2, 3.4]
+
+_cut_displayed: ClassVar[str] = "yz"
+_default_figsize: ClassVar[list[float, float]] = (2.2, 3.4)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_projectors.py:507]:
```suggestion
_default_figsize: ClassVar[list[float, float]] = [2.2, 3.4]
```

File: nilearn/plotting/displays/_projectors.py [hunk 18]
```diff
@@ -700,8 +705,9 @@ class LProjector(OrthoProjector):
 
     """
 
-    _cut_displayed = "l"
-    _default_figsize = [2.6, 3.0]
+
+_cut_displayed: ClassVar[str] = "l"
+_default_figsize: ClassVar[list[float, float]] = (2.6, 3.0)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_projectors.py:710]:
```suggestion
_default_figsize: ClassVar[list[float, float]] = [2.6, 3.0]
```

File: nilearn/plotting/displays/_projectors.py [hunk 19]
```diff
@@ -733,8 +739,9 @@ class RProjector(OrthoProjector):
 
     """
 
-    _cut_displayed = "r"
-    _default_figsize = [2.6, 2.8]
+
+_cut_displayed: ClassVar[str] = "r"
+_default_figsize: ClassVar[list[float, float]] = (2.6, 2.8)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_projectors.py:744]:
```suggestion
_default_figsize: ClassVar[list[float, float]] = [2.6, 2.8]
```

File: nilearn/plotting/displays/_slicers.py [hunk 20]
```diff
@@ -46,7 +47,7 @@ class BaseSlicer:
     """
 
     # This actually encodes the figsize for only one axe
-    _default_figsize = [2.2, 2.6]
+    _default_figsize: ClassVar[list[float, float]] = (2.2, 2.6)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_slicers.py:50]:
```suggestion
    _default_figsize: ClassVar[list[float, float]] = [2.2, 2.6]
```

File: nilearn/plotting/displays/_slicers.py [hunk 21]
```diff
@@ -895,9 +896,9 @@ class OrthoSlicer(BaseSlicer):
 
     """
 
-    _cut_displayed = "yxz"
+    _cut_displayed: ClassVar[str] = "yxz"
     _axes_class = CutAxes
-    _default_figsize = [2.2, 3.5]
+    _default_figsize: ClassVar[list[float, float]] = (2.2, 3.5)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_slicers.py:901]:
```suggestion
    _default_figsize: ClassVar[list[float, float]] = [2.2, 3.5]
```

File: nilearn/plotting/displays/_slicers.py [hunk 22]
```diff
@@ -1124,9 +1125,9 @@ class TiledSlicer(BaseSlicer):
 
     """
 
-    _cut_displayed = "yxz"
+    _cut_displayed: ClassVar[str] = "yxz"
     _axes_class = CutAxes
-    _default_figsize = [2.0, 7.6]
+    _default_figsize: ClassVar[list[float, float]] = (2.0, 7.6)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_slicers.py:1130]:
```suggestion
    _default_figsize: ClassVar[list[float, float]] = [2.0, 7.6]
```

File: nilearn/plotting/displays/_slicers.py [hunk 23]
```diff
@@ -1636,8 +1637,8 @@ class XSlicer(BaseStackedSlicer):
 
     """
 
-    _direction = "x"
-    _default_figsize = [2.6, 2.3]
+    _direction: ClassVar[str] = "x"
+    _default_figsize: ClassVar[list[float, float]] = (2.6, 2.3)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_slicers.py:1641]:
```suggestion
    _default_figsize: ClassVar[list[float, float]] = [2.6, 2.3]
```

File: nilearn/plotting/displays/_slicers.py [hunk 24]
```diff
@@ -1675,8 +1676,8 @@ class YSlicer(BaseStackedSlicer):
 
     """
 
-    _direction = "y"
-    _default_figsize = [2.2, 3.0]
+    _direction: ClassVar[str] = "y"
+    _default_figsize: ClassVar[list[float, float]] = (2.2, 3.0)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_slicers.py:1680]:
```suggestion
    _default_figsize: ClassVar[list[float, float]] = [2.2, 3.0]
```

File: nilearn/plotting/displays/_slicers.py [hunk 25]
```diff
@@ -1714,8 +1715,8 @@ class ZSlicer(BaseStackedSlicer):
 
     """
 
-    _direction = "z"
-    _default_figsize = [2.2, 3.2]
+    _direction: ClassVar[str] = "z"
+    _default_figsize: ClassVar[list[float, float]] = (2.2, 3.2)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_slicers.py:1719]:
```suggestion
    _default_figsize: ClassVar[list[float, float]] = [2.2, 3.2]
```

File: nilearn/plotting/displays/_slicers.py [hunk 26]
```diff
@@ -1829,8 +1830,8 @@ class YZSlicer(OrthoSlicer):
 
     """
 
-    _cut_displayed = "yz"
-    _default_figsize = [2.2, 3.0]
+    _cut_displayed: ClassVar[str] = "yz"
+    _default_figsize: ClassVar[list[float, float]] = (2.2, 3.0)
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_slicers.py:1834]:
```suggestion
    _default_figsize: ClassVar[list[float, float]] = [2.2, 3.0]
```

File: nilearn/plotting/displays/_projectors.py [hunk 27]
```diff
@@ -390,8 +391,9 @@ class ZProjector(OrthoProjector):
 
     """
 
-    _cut_displayed = "z"
-    _default_figsize = [2.2, 3.4]
+
+_cut_displayed: ClassVar[str] = "z"
+_default_figsize: ClassVar[list[float, float]] = [2.2, 3.4]
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_projectors.py:396]:
```suggestion
    _cut_displayed: ClassVar[str] = "z"
    _default_figsize: ClassVar[list[float, float]] = [2.2, 3.4]
```

File: nilearn/plotting/displays/_projectors.py [hunk 28]
```diff
@@ -498,8 +500,9 @@ class YZProjector(OrthoProjector):
 
     """
 
-    _cut_displayed = "yz"
-    _default_figsize = [2.2, 3.4]
+
+_cut_displayed: ClassVar[str] = "yz"
+_default_figsize: ClassVar[list[float, float]] = [2.2, 3.4]
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_projectors.py:505]:
```suggestion
    _cut_displayed: ClassVar[str] = "yz"
    _default_figsize: ClassVar[list[float, float]] = [2.2, 3.4]
```

File: nilearn/plotting/displays/_projectors.py [hunk 29]
```diff
@@ -700,8 +703,9 @@ class LProjector(OrthoProjector):
 
     """
 
-    _cut_displayed = "l"
-    _default_figsize = [2.6, 3.0]
+
+_cut_displayed: ClassVar[str] = "l"
+_default_figsize: ClassVar[list[float, float]] = [2.6, 3.0]
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_projectors.py:708]:
```suggestion
    _cut_displayed: ClassVar[str] = "l"
    _default_figsize: ClassVar[list[float, float]] = [2.6, 3.0]
```

File: nilearn/plotting/displays/_projectors.py [hunk 30]
```diff
@@ -733,8 +737,9 @@ class RProjector(OrthoProjector):
 
     """
 
-    _cut_displayed = "r"
-    _default_figsize = [2.6, 2.8]
+
+_cut_displayed: ClassVar[str] = "r"
+_default_figsize: ClassVar[list[float, float]] = [2.6, 2.8]
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_projectors.py:742]:
```suggestion
    _cut_displayed: ClassVar[str] = "r"
    _default_figsize: ClassVar[list[float, float]] = [2.6, 2.8]
```

File: doc/changes/latest.rst [hunk 31]
```diff
@@ -16,6 +16,7 @@ Fixes
 
 - :bdg-dark:`Code` Ensure that low alpha values in :func:`~plotting.plot_img_on_surf` do not cause color values to be above 1 (:gh:`4906` by `Paul Bogdan`_).
 
+- :bdg-dark:`Code` Fix all occurrences of the RUF012 error related to mutable default values in Python classes (:gh:`4954` by `Idrissa Traore_).
```
Comment (Remi-Gau) [inline doc/changes/latest.rst:19]:
```suggestion
- :bdg-dark:`Code` Fix all occurrences of the RUF012 error related to mutable default values in Python classes (:gh:`4954` by `Idrissa Traore`_).
```

**Review by bthirion — APPROVED** [review]
Thx !

**Pull request merged**